dnf5-plugins/automatic_plugin/automatic.cpp

```diff
@@ -0,0 +1,468 @@
+#include <libdnf5/utils/bgettext/bgettext-mark-domain.h>
+#include <libdnf5/utils/format.hpp>
+#include <libsmartcols/libsmartcols.h>
+#include <netdb.h>
+#include <stdlib.h>
+
+#include <chrono>
+#include <filesystem>
+#include <iostream>
+#include <random>
+#include <set>
+#include <string>
+#include <thread>
+
+namespace {
+
+/// Sleep for random number of seconds in interval <0, max_value>
+static void random_wait(int max_value) {
+    std::random_device rd;
+    std::mt19937 rng(rd());
+    std::uniform_int_distribution<int> distribution(0, max_value);
+
+    sleep(distribution(rng));
+}
+
+static bool reboot_needed(const libdnf5::base::Transaction & transaction) {
+    static const std::set<std::string> need_reboot = {
+        "kernel", "kernel-rt", "glibc", "linux-firmware", "systemd", "dbus", "dbus-broker", "dbus-daemon"};
+    for (const auto & pkg : transaction.get_transaction_packages()) {
+        if (need_reboot.find(pkg.get_package().get_name()) != need_reboot.end()) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static bool server_available(std::string_view host, std::string_view service) {
+    // Resolve host name/service to IP address/port
+    struct addrinfo * server_info;
+    if (getaddrinfo(host.data(), service.data(), nullptr, &server_info) != 0) {
+        return false;
+    }
+
+    // Create a socket
+    int sockfd = socket(AF_INET, SOCK_STREAM, 0);
+    if (sockfd < 0) {
+        freeaddrinfo(server_info);
+        return false;
+    }
+
+    // Attempt to connect to the server
+    bool retval = true;
+    struct timeval timeout;
+    timeout.tv_sec = 1;
+    timeout.tv_usec = 0;
+    setsockopt(sockfd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
+    if (connect(sockfd, server_info->ai_addr, server_info->ai_addrlen) < 0) {
+        retval = false;
+    }
+
+    close(sockfd);
+    freeaddrinfo(server_info);
+    return retval;
+}
+
+}  // namespace
+
+
+namespace dnf5 {
+
+void AutomaticCommand::wait_for_network() {
+    auto timeout = config_automatic.config_commands.network_online_timeout.get_value();
+    if (timeout <= 0) {
+        return;
+    }
+
+    auto & context = get_context();
+    auto & base = context.base;
+    auto & logger = *base.get_logger();
+    logger.debug("Waiting for internet connection...");
+
+    const auto time_0 = std::chrono::system_clock::now();
+    const auto time_end = time_0 + std::chrono::seconds(timeout);
+
+    // collect repository addresses to check network availability on
+    libdnf5::repo::RepoQuery repos(base);
+    repos.filter_enabled(true);
+    std::vector<std::string> urls{};
+    for (const auto & repo : repos) {
+        std::string proxy{};
+        try {
+            proxy = repo->get_config().get_proxy_option().get_value();
+            if (!proxy.empty()) {
+                urls.emplace_back(proxy);
+                continue;
+            }
+        } catch (const libdnf5::OptionValueNotSetError & e) {
```

Comment:
Only note: It looks strange that you have to handle `libdnf5::OptionValueNotSetError` and `(!proxy.empty())`